x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/entity_store/routes/entity_crud/upsert_entity.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { buildRouteValidationWithZod } from '@kbn/zod-helpers';
+import type { IKibanaResponse, Logger } from '@kbn/core/server';
+import type { UpsertEntityResponse } from '../../../../../../common/api/entity_analytics/entity_store/entities/upsert_entity.gen';
+import {
+  UpsertEntityRequestBody,
+  UpsertEntityRequestParams,
+  UpsertEntityRequestQuery,
+} from '../../../../../../common/api/entity_analytics/entity_store/entities/upsert_entity.gen';
+import type { EntityAnalyticsRoutesDeps } from '../../../types';
+import { API_VERSIONS, APP_ID } from '../../../../../../common/constants';
+import { BadCRUDRequestError, DocumentNotFoundError, EngineNotRunningError } from '../../errors';
+import { CapabilityNotEnabledError } from '../../errors/capability_not_enabled_error';
+
+export const upsertEntity = (router: EntityAnalyticsRoutesDeps['router'], logger: Logger) => {
+  router.versioned
+    .put({
+      access: 'public',
+      path: '/api/entity_store/entities/{entityType}',
+      security: {
+        authz: {
+          requiredPrivileges: ['securitySolution', `${APP_ID}-entity-analytics`],
+        },
+      },
+    })
+    .addVersion(
+      {
+        version: API_VERSIONS.public.v1,
+        validate: {
+          request: {
+            query: buildRouteValidationWithZod(UpsertEntityRequestQuery),
+            params: buildRouteValidationWithZod(UpsertEntityRequestParams),
+            body: buildRouteValidationWithZod(UpsertEntityRequestBody),
+          },
+        },
+      },
+      async (context, request, response): Promise<IKibanaResponse<UpsertEntityResponse>> => {
+        const secSol = await context.securitySolution;
+
+        try {
+          await secSol
+            .getEntityStoreCrudClient()
+            .upsertEntity(request.params.entityType, request.body, request.query.force);
+
+          return response.ok();
+        } catch (error) {
+          if (
+            error instanceof EngineNotRunningError ||
+            error instanceof CapabilityNotEnabledError
+          ) {
+            // Service Unavailable 503
+            return response.customError({ statusCode: 503, body: error as EngineNotRunningError });
+          }
+
+          if (error instanceof BadCRUDRequestError) {
+            return response.badRequest({ body: error as BadCRUDRequestError });
+          }
+
+          if (error instanceof DocumentNotFoundError) {
+            return response.notFound({ body: error as DocumentNotFoundError });
+          }
+
+          logger.error(error);
```

Comment:
Maybe log the error first? before all the cases...

Reply:
Does it make sense to log before known errors? My thought through is more like, not found is not a trouble shooting error, not enabled is not a trouble shooting error, they expected errors.

I log those who deviate of that so we can trouble shoot eventually. 

Curious to hear your opinion :)